### PR TITLE
App spec integration into cocoapods

### PR DESCRIPTION
**Pull request by dostrander**
@dnkoutso @segiddins

### CODE REVIEW

**Comment by CocoaPodsBarista** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please include a CHANGELOG entry to credit yourself! <br />
You can find it at <a href="https://github.com/CocoaPods/CocoaPods/blob/master/CHANGELOG.md">CHANGELOG.md</a>.</td>
    </tr>
  </tbody>
</table>

Here's an example of your CHANGELOG entry:

```markdown
* App spec integration into cocoapods  
  [dostrander](https://github.com/dostrander)
  [#issue_number](https://github.com/CocoaPods/CocoaPods/issues/issue_number)
```

*note*: There are two invisible spaces after the entry's text.


<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

File: lib/cocoapods/target/pod_target.rb [hunk 1]
```diff
@@ -356,6 +381,16 @@ def test_target_label(test_spec)
       "#{label}-#{test_spec.test_type.capitalize}-#{test_spec.name.split('/')[1..-1].join('-')}"
     end
 
+    # @param  [Specification] app_spec
+    #         The app spec to use for producing the test label.
+    #
+    # @return [String] The derived name of the test target.
+    #
+    def app_target_label(app_spec)
+      ## TODO: THIS NEEDS TO BE UPDATED TO SOMETHING
```
Comment (segiddins) [inline lib/cocoapods/target/pod_target.rb]: TODO

File: spec/functional/command/init_spec.rb [hunk 2]
```diff
@@ -102,8 +102,8 @@ module Pod
           # platform :ios, '9.0'
 
           target 'App' do
-            # Uncomment the next line if you're using Swift or would like to use dynamic frameworks
-            # use_frameworks!
+            # Comment the next line if you're not using Swift and don't want to use dynamic frameworks
```
Comment (segiddins) [inline spec/functional/command/init_spec.rb]: why this change?

File: spec/unit/installer/analyzer/pod_variant_set_spec.rb [hunk 3]
```diff
@@ -14,83 +14,83 @@ module Pod
       PodVariant = Pod::Installer::Analyzer::PodVariant.freeze
 
       it 'returns an empty scope if there is only one variant' do
-        variants = PodVariantSet.new([PodVariant.new([@root_spec], [], Platform.ios)])
+        variants = PodVariantSet.new([PodVariant.new([@root_spec], [], [], Platform.ios)])
```
Comment (segiddins) [inline spec/unit/installer/analyzer/pod_variant_set_spec.rb]: I think we'll want a test case that covers the new array being non-empty

File: spec/unit/installer/analyzer/pod_variant_spec.rb [hunk 4]
```diff
@@ -11,46 +11,46 @@ class Analyzer
...
           v0 = stub('Value at index 0')
-          k1 = PodVariant.new(@specs, [], @platform, true)
+          k1 = PodVariant.new(@specs, [], [], @platform, true)
           v1 = stub('Value at index 1')
           hash = { k0 => v0, k1 => v1 }
           hash[k0].should == v0
           hash[k1].should == v1
         end
 
         it 'does not use testspecs for equality' do
```
Comment (segiddins) [inline spec/unit/installer/analyzer/pod_variant_spec.rb]: we also need a test that app specs are used/not used for hashing/equality

File: spec/unit/installer/analyzer/target_inspector_spec.rb [hunk 5]
```diff
@@ -303,7 +303,7 @@ module Pod
         target_inspector.send(:compute_swift_version_from_targets, user_targets).should.equal '2.3'
       end
 
-      it 'returns nil if the version is not defined' do
+      it 'returns default if the version is not defined' do
```
Comment (segiddins) [inline spec/unit/installer/analyzer/target_inspector_spec.rb]: why did this change

File: spec/functional/command/init_spec.rb (hunk 2 above)
Comment (dostrander) [reply spec/functional/command/init_spec.rb]: From the changes i picked up [here](https://github.com/CocoaPods/Xcodeproj/pull/611) had to update the tests

File: lib/cocoapods/installer/analyzer.rb [hunk 6]
```diff
@@ -560,18 +566,21 @@ def generate_pod_targets(resolver_specs_by_target, target_inspections)
...
-              test_specs, specs = all_specs.partition(&:test_specification?)
-              pod_variant = PodVariant.new(specs, test_specs, target_definition.platform, target_definition.uses_frameworks?)
+              library_specs = all_specs.select(&:library_specification?)
+              test_specs = all_specs.select(&:test_specification?)
+              app_specs = all_specs.select(&:app_specification?)
+              pod_variant = PodVariant.new(library_specs, test_specs, app_specs, target_definition.platform, target_definition.uses_frameworks?)
               hash[root_spec] ||= {}
               (hash[root_spec][pod_variant] ||= []) << target_definition
               hash[root_spec].keys.find { |k| k == pod_variant }.test_specs.concat(test_specs).uniq!
+              hash[root_spec].keys.find { |k| k == pod_variant }.app_specs.concat(app_specs).uniq!
```
[the remark this answers is not in the document]
Comment (dostrander) [reply lib/cocoapods/installer/analyzer.rb]: 👍

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 7]
```diff
@@ -92,6 +92,10 @@ def write
               installation_result.test_native_targets.each do |test_native_target|
                 scheme.add_test_target(test_native_target)
               end
+
+              installation_result.app_native_targets.each do |add_build_target|
```
[the remark this answers is not in the document]
Comment (dostrander) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: That's actually a good question. As of the current state of cocoapods with PR my thinking is no because of the fact there is still a target that will be defined as the main "app". 

I don't feel too strongly about that though and could be easily convinced

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 8]
```diff
@@ -518,6 +568,82 @@ def create_test_target_embed_frameworks_script(test_spec)
...
+          #
+          # @param  [Hash{String=>Array<PBXNativeTarget>}] app_resource_bundle_targets
+          #         the additional app resource bundle targets to link the xcconfig file into.
+          #
+          # @return [void]
+          #
+          def create_app_xcconfig_files(app_native_targets, app_resource_bundle_targets)
+            target.app_specs.each do |app_spec|
+              spec_consumer = app_spec.consumer(target.platform)
+              path = target.xcconfig_path(app_spec.name.split('/')[1..-1].join('-').to_s)
```
[the remark this answers is not in the document]
Comment (dostrander) [reply lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: I'm tempted to put this on the `specification.rb` itself and call it `subspec_name`. It's used in the codebase ~5 times. LMKWYT

File: lib/cocoapods/resolver/resolver_specification.rb [hunk 9]
```diff
@@ -17,9 +17,15 @@ class ResolverSpecification
       attr_reader :used_by_tests_only
       alias used_by_tests_only? used_by_tests_only
 
-      def initialize(spec, used_by_tests_only, source)
+      # @return [Bool] whether this resolved specification is only used by apps.
+      #
+      attr_reader :used_by_apps_only
```
[the remark this answers is not in the document]
Comment (dostrander) [reply lib/cocoapods/resolver/resolver_specification.rb]: Good idea 👍

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb (hunk 8 above)
[the remark this answers is not in the document]
Comment (dostrander) [reply lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: Actually ended up putting it on the target because it's all used within the target or target installer.

File: Gemfile [hunk 10]
```diff
@@ -25,7 +25,9 @@ gem 'json', :git => 'https://github.com/segiddins/json.git', :branch => 'seg-1.7
 
 group :development do
   cp_gem 'claide',                'CLAide'
-  cp_gem 'cocoapods-core',        'Core'
+  # DO NOT MERGE!
```
[the remark this answers is not in the document]
Comment (dostrander) [reply Gemfile]: Removed, just need to merge the core changes for these to pass

File: lib/cocoapods/installer/analyzer.rb [hunk 11]
```diff
@@ -583,13 +593,19 @@ def generate_pod_targets(resolver_specs_by_target, target_inspections)
...
-            dependencies = dependencies_for_specs(target.non_test_specs.to_set, target.platform, all_specs.dup).group_by(&:root)
+            dependencies = dependencies_for_specs(target.library_specs.to_set, target.platform, all_specs.dup).group_by(&:root)
             target.dependent_targets = filter_dependencies(dependencies, pod_targets_by_name, target)
             target.test_dependent_targets_by_spec_name = target.test_specs.each_with_object({}) do |test_spec, hash|
               test_dependencies = dependencies_for_specs([test_spec], target.platform, all_specs).group_by(&:root)
               test_dependencies.delete_if { |k| dependencies.key? k }
               hash[test_spec.name] = filter_dependencies(test_dependencies, pod_targets_by_name, target)
             end
+
+            target.app_dependent_targets_by_spec_name = target.app_specs.each_with_object({}) do |app_spec, hash|
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer.rb]: would love to consolidate this into a single method since it does the same for test specs.

File: lib/cocoapods/installer/analyzer/pod_variant.rb [hunk 12]
```diff
@@ -30,12 +34,14 @@ def root_spec
         #
         # @param [Array<Specification>] specs      @see #specs
         # @param [Array<Specification>] test_specs @see #test_specs
+        # @param [Array<Specification>] app_specs  @see #app_specs
         # @param [Platform] platform               @see #platform
         # @param [Bool] requires_frameworks        @see #requires_frameworks?
         #
-        def initialize(specs, test_specs, platform, requires_frameworks = false)
+        def initialize(specs, test_specs, app_specs, platform, requires_frameworks = false)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/analyzer/pod_variant.rb]: can you please update doc in line 50 about test specs + app specs not being considered for equality `==`?

File: lib/cocoapods/installer/xcode/pods_project_generator.rb [hunk 13]
```diff
@@ -299,6 +311,24 @@ def wire_target_dependencies(target_installation_results)
...
+                app_dependent_targets.each do |app_dependent_target|
+                  dependency_installation_result = pod_target_installation_results_hash[app_dependent_target.name]
+                  resource_bundle_native_targets = dependency_installation_result.app_resource_bundle_targets[app_specs.first.name]
+                  unless resource_bundle_native_targets.nil?
+                    resource_bundle_native_targets.each do |app_resource_bundle_target|
+                      app_native_target.add_dependency(app_resource_bundle_target)
+                    end
+                  end
+                  app_native_target.add_dependency(dependency_installation_result.native_target)
+                  add_framework_file_reference_to_native_target(app_native_target, pod_target, app_dependent_target, frameworks_group)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator.rb]: is this needed for app targets?

File: lib/cocoapods/installer/xcode/pods_project_generator/app_host_installer.rb [hunk 14]
```diff
@@ -21,47 +21,53 @@ class AppHostInstaller
...
           # @param [Pod::Project] project @see #project
           # @param [Platform] platform @see #platform
-          # @param [String] name @see #name
-          # @param [String] pod_name @see #pod_name
+          # @param [String] subspec_name @see #subspec_name
+          # @param [String] spec_name @see #spec_name
+          # @param [String] app_target_label see #app_target_label
           #
-          def initialize(sandbox, project, platform, name, pod_name)
+          def initialize(sandbox, project, platform, subspec_name, spec_name, app_target_label)
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/app_host_installer.rb]: `subspec_name` is awfully specific for the needs of a label here. Is there a better name for this? I do not quite understand the difference of passing a `spec_name` and `subspec_name` easily.

File: lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb [hunk 15]
```diff
@@ -348,6 +366,57 @@ def add_test_app_host_targets(test_native_targets)
...
+
+              target.user_build_configurations.each do |bc_name, type|
+                app_native_target.add_build_configuration(bc_name, type)
+              end
+
+              app_native_target.build_configurations.each do |configuration|
+                configuration.build_settings.merge!(custom_build_settings)
+
+                # target_installer will automatically add an empty `OTHER_LDFLAGS`. For app
+                # targets those are set via a test xcconfig file instead.
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/pod_target_installer.rb]: nit `test xcconfig`

File: lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb [hunk 16]
```diff
@@ -38,6 +38,18 @@ class TargetInstallationResult
           #
           attr_reader :test_app_host_targets
 
+          # @return [Array<PBXNativeTarget>] app_native_targets
+          #         The app native targets that were produced for this target. Can be empty if there were no app
+          #         native targets created (e.g. no test specs present).
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/target_installation_result.rb]: "test specs" -> "app specs"

File: lib/cocoapods/target/build_settings.rb [hunk 17]
```diff
@@ -814,12 +838,12 @@ def pod_target_xcconfig_values_by_consumer_by_key
 
         # @return [Array<Sandbox::FileAccessor>]
         define_build_settings_method :file_accessors, :memoized => true do
-          target.file_accessors.select { |fa| fa.spec.test_specification? == test_xcconfig? }
+          target.file_accessors.select { |c| c.spec.spec_type == @xcconfig_spec_type }
```
Comment (dnkoutso) [inline lib/cocoapods/target/build_settings.rb]: why rename `fa` to `c`?

File: lib/cocoapods/target/pod_target.rb [hunk 18]
```diff
@@ -347,13 +373,31 @@ def resources_bundle_target_label(bundle_name)
       "#{label}-#{bundle_name}"
     end
 
+    # @param  [Specification] subspec
+    #         The subspec to use for producing the label.
+    #
+    # @return [String] The derived name of the test target.
+    #
+    def subspec_label(subspec)
+      subspec.name.split('/')[1..-1].join('-').to_s
```
Comment (dnkoutso) [inline lib/cocoapods/target/pod_target.rb]: probably `raise `if subspec.root?` to guard?

File: spec/unit/installer/xcode/pods_project_generator/file_references_installer_spec.rb [hunk 19]
```diff
@@ -208,15 +208,21 @@ class PodsProjectGenerator
           describe 'Private Helpers' do
             describe '#file_accessors' do
               it 'returns the file accessors' do
-                pod_target_1 = PodTarget.new(config.sandbox, false, {}, [], Platform.ios, [stub('Spec', :test_specification? => false)], [fixture_target_definition], [fixture_file_accessor('banana-lib/BananaLib.podspec')])
-                pod_target_2 = PodTarget.new(config.sandbox, false, {}, [], Platform.ios, [stub('Spec', :test_specification? => false)], [fixture_target_definition], [fixture_file_accessor('banana-lib/BananaLib.podspec')])
+                pod_target_1 = PodTarget.new(config.sandbox, false, {}, [], Platform.ios,
```
Comment (dnkoutso) [inline spec/unit/installer/xcode/pods_project_generator/file_references_installer_spec.rb]: can maybe use `fixture_pod_target` method here.

File: lib/cocoapods/target/build_settings.rb (hunk 17 above)
Comment (dostrander) [reply lib/cocoapods/target/build_settings.rb]: I think this had a couple revs on it and just happened to land there, i'll change back

File: lib/cocoapods/installer/xcode/pods_project_generator.rb (hunk 13 above)
Comment (dnkoutso) [reply lib/cocoapods/installer/xcode/pods_project_generator.rb]: talked offline yes its needed

File: lib/cocoapods/installer/xcode/pods_project_generator/app_host_installer.rb [hunk 20]
```diff
@@ -21,47 +21,53 @@ class AppHostInstaller
           #
           attr_reader :platform
 
-          # @return [String] the name to use for this app host target.
+          # @return [String] the name of the sub group.
           #
-          attr_reader :name
+          attr_reader :subgroup_name
```
Comment (dnkoutso) [inline lib/cocoapods/installer/xcode/pods_project_generator/app_host_installer.rb]: much better name imo

**Comment by dnkoutso** [comment]
There is a lot of code duplication in here which I am generally not a fan of. Before 1.7.0 we should attempt to reduce a bit the duplication if possible.

Merging.

**Comment by revolter** [comment]
What's an "app spec"?

**Comment by orta** [comment]
http://blog.cocoapods.org/CocoaPods-1.7.0-beta/
